third-party/CMakeLists.txt

```diff
@@ -0,0 +1,226 @@
+cmake_host_system_information(RESULT available_memory_mb QUERY AVAILABLE_PHYSICAL_MEMORY)
+
+execute_process(
+    COMMAND ldd --version
+    COMMAND head -1
+    COMMAND cut -d ")" -f 2
+    COMMAND cut -d " " -f 2
+    OUTPUT_VARIABLE GLIBC_VERSION
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+)
+
+# Guess the number of building jobs based on the available memories
+set(jobs_by_cpu ${num_cores})
+math(EXPR jobs_by_mem "${available_memory_mb} / 1024 / 2")
+if (jobs_by_mem EQUAL 0)
+    set(jobs_by_mem 1)
+endif()
+
+# Set the number of building jobs to min(jobs_by_mem, jobs_by_cpu),
+# iff BUILDING_JOBS_NUM has not been set or set to 0.
+if (NOT BUILDING_JOBS_NUM OR BUILDING_JOBS_NUM EQUAL 0)
+    set(BUILDING_JOBS_NUM ${jobs_by_cpu})
+    if (BUILDING_JOBS_NUM GREATER jobs_by_mem)
+        set(BUILDING_JOBS_NUM ${jobs_by_mem})
+    endif()
+endif()
+
+message(STATUS "Number of online physcial CPU cores: ${num_cores}")
+message(STATUS "Available physical memory: ${available_memory_mb} MB")
+message(STATUS "Building third party with ${BUILDING_JOBS_NUM} jobs")
+message(STATUS "Glibc version: ${GLIBC_VERSION}")
+
+if (NOT DOWNLOAD_DIR)
+    set(DOWNLOAD_DIR ${CMAKE_CURRENT_BINARY_DIR}/downloads)
+endif()
+set(BUILD_INFO_DIR ${CMAKE_CURRENT_BINARY_DIR}/build-info)
+set(BUILDING_PATH "${CMAKE_INSTALL_PREFIX}/bin:${CMAKE_INSTALL_PREFIX}/sbin:$ENV{PATH}")
+set(ACLOCAL_PATH "${CMAKE_INSTALL_PREFIX}/share/aclocal:${CMAKE_INSTALL_PREFIX}/share/aclocal-1.15")
+
+set(extra_link_libs "-static-libstdc++ -static-libgcc")
+
+if(GLIBC_VERSION VERSION_LESS 2.17)
+    set(extra_link_libs "${extra_link_libs} -lrt")
+endif()
+
+set(common_cmake_args
+    -DCMAKE_INSTALL_PREFIX=${CMAKE_INSTALL_PREFIX}
+    -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
+    -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+    "-DCMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS} -fPIC"
+    "-DCMAKE_C_FLAGS=${CMAKE_C_FLAGS} -fPIC"
+    "-DCMAKE_EXE_LINKER_FLAGS=${extra_link_libs}"
+    "-DCMAKE_SHARED_LINKER_FLAGS=${extra_link_libs}"
+    -DCMAKE_INCLUDE_PATH=${CMAKE_INSTALL_PREFIX}/include
+    -DCMAKE_LIBRARY_PATH=${CMAKE_INSTALL_PREFIX}/lib
+)
+
+set(common_configure_args
+    --prefix=${CMAKE_INSTALL_PREFIX}
+    --quiet
+)
+
+set(common_configure_envs
+    "env"
+    "CC=${CMAKE_C_COMPILER}"
+    "CXX=${CMAKE_CXX_COMPILER}"
+    "CFLAGS=${CMAKE_C_FLAGS} -fPIC -O2 -D_DEFAULT_SOURCE -D_GNU_SOURCE"
+    "CXXFLAGS=${CMAKE_CXX_FLAGS} -fPIC -O2 -D_DEFAULT_SOURCE -D_GNU_SOURCE"
+    "CPPFLAGS=-isystem ${CMAKE_INSTALL_PREFIX}/include"
+    "LDFLAGS=-L${CMAKE_INSTALL_PREFIX}/lib -L${CMAKE_INSTALL_PREFIX}/lib64"
+    "PATH=${BUILDING_PATH}"
+    "ACLOCAL_PATH=${ACLOCAL_PATH}"
+)
+
+set(ALL_TARGETS
+    libunwind
+    openssl
+    libevent
+    keyutils
+    krb5
+    jemalloc
+    bzip2
+    zstd
+    zlib
+    snappy
+    lzma
+    lz4
+    libaio
+    mstch
+    fatal
+    double-conversion
+    gflags
+    googletest
+    glog
+    boost
+    rocksdb
+    folly
+    wangle
+    proxygen
+    fbthrift
+    capstone
```

Comment:
interesting.

Reply:
capstone is an instruction disassembler. I wrote a tool to analyze instruction set a binary uses, which I haven't post it out.